### PR TITLE
KOGITO-2421 Add Milestones to runtime events

### PR DESCRIPTION
**Pull request by cristianonicolai**
https://issues.redhat.com/browse/KOGITO-2421

Include information about milestones, in cloud events produced by the core engine.

### CODE REVIEW

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 1]
```diff
@@ -56,14 +56,14 @@ void testSimpleMilestone() throws Exception {
         assertState(processInstance, ProcessInstance.STATE_PENDING);
 
         Collection<Milestone> expected = new ArrayList<>();
-        expected.add(new Milestone.Builder("").withName("AutoStartMilestone").withStatus(AVAILABLE).build());
-        expected.add(new Milestone.Builder("").withName("SimpleMilestone").withStatus(AVAILABLE).build());
+        expected.add(Milestone.builder().withId("").withName("AutoStartMilestone").withStatus(AVAILABLE).build());
```
Comment (ruromero) [inline kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:59]: then why set the id?

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 2]
```diff
@@ -56,14 +56,14 @@ void testSimpleMilestone() throws Exception {
...
+        expected.add(Milestone.builder().withId("").withName("SimpleMilestone").withStatus(AVAILABLE).build());
```
Comment (ruromero) [inline kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:60]: no need to set the id there

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 3]
```diff
@@ -91,20 +91,20 @@ void testConditionalMilestone() throws Exception {
         assertState(processInstance, ProcessInstance.STATE_PENDING);
 
         Collection<Milestone> expected = new ArrayList<>();
-        expected.add(new Milestone.Builder("").withName("Milestone").withStatus(AVAILABLE).build());
+        expected.add(Milestone.builder().withId("").withName("Milestone").withStatus(AVAILABLE).build());
```
Comment (ruromero) [inline kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:94]: same

File: kogito-codegen/src/test/java/org/kie/kogito/codegen/tests/PublishEventTest.java [hunk 4]
```diff
@@ -40,15 +38,59 @@
...
+        assertThat(processDataEvent.getKogitoProcessinstanceId()).isNotNull();
+        assertThat(processDataEvent.getKogitoParentProcessinstanceId()).isNull();
+        assertThat(processDataEvent.getKogitoRootProcessinstanceId()).isNull();
+        assertThat(processDataEvent.getKogitoProcessId()).isEqualTo("TestCase.SimpleMilestone");
+        assertThat(processDataEvent.getKogitoProcessinstanceState()).isEqualTo("2");
+        assertThat(processDataEvent.getSource()).isEqualTo("http://myhost/SimpleMilestone");
+
+        Set<MilestoneEventBody> milestones = ((ProcessInstanceDataEvent) event).getData().getMilestones();
+        assertThat(milestones).hasSize(2).containsExactly(
+                MilestoneEventBody.create().id("_18950AC4-664B-465E-B99E-766D38D362BF").name("AutoStartMilestone").status(Status.COMPLETED.name()).build(),
```
Comment (ruromero) [inline kogito-codegen/src/test/java/org/kie/kogito/codegen/tests/PublishEventTest.java:90]: Try to avoid using the hardcoded node ids

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java [hunk 5]
```diff
@@ -391,63 +369,12 @@ public void transitionWorkItem(String id, Transition<?> transition) {
...
-    private Status getStatus(String uid) {
-        RuleFlowProcessInstance processInstance = ((RuleFlowProcessInstance) legacyProcessInstance);
-        if (processInstance.getCompletedNodeIds().contains(uid)) {
-            return COMPLETED;
-        }
-        if (processInstance.getActiveNodeIds().contains(uid)) {
-            return ACTIVE;
-        }
-        return AVAILABLE;
+        return ((WorkflowProcessInstance) legacyProcessInstance).adHocFragments();
```
Comment (ruromero) [inline jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:377]: Why did you move it to the legacyProcessInstance? Shouldn't we avoid adding more functionality there?

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java [hunk 6]
```diff
@@ -391,63 +369,12 @@ public void transitionWorkItem(String id, Transition<?> transition) {
...
-        return getNodes(getProcessNodeContainer(), MilestoneNode.class)
-                .map(n -> {
-                    String uid = (String) n.getMetaData().get(UNIQUE_ID);
-                    return new Milestone.Builder(uid)
-                            .withName(n.getName())
-                            .withStatus(getStatus(uid))
-                            .build();
-                })
-                .collect(Collectors.toList());
+        return ((WorkflowProcessInstance) legacyProcessInstance).milestones();
```
Comment (ruromero) [inline jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:372]: If the motivation of moving the methods to the legacyProcessInstance is because Events contains a ProcessInstance implementation (i.e. workflowProcessInstance) wouldn't it be more appropriate to have the events return a ProcessInstance<T> meaning an AbstractProcessInstance? Based on the name of this attribute (`legacyProcessInstance`) I'd like to avoid adding more functionality.

I haven't really looked into the implications of that, just asking your opinion.

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java [hunk 7]
```diff
@@ -16,21 +16,12 @@
...
-import java.util.Collections;
-import java.util.Date;
-import java.util.HashMap;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (ruromero) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java:19]: Can you update your IDE preferences?

File: api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/MilestoneEventBody.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            if (other.id != null) {
+                return false;
+            }
+        } else if (!id.equals(other.id)) {
+            return false;
+        }
+        return true;
+    }
+
+    public Builder update() {
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/MilestoneEventBody.java:82]: Is there so it allows updating the event as there are no setter methods.

File: api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBody.java [hunk 9]
```diff
@@ -52,6 +52,8 @@
     private ProcessErrorEventBody error;
     
     private List<String> roles;
+
+    private Set<MilestoneEventBody> milestones = new LinkedHashSet<>();
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBody.java:56]: will replace with an empty set as this is overridden anyway.

File: api/kogito-services/src/test/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBatchTest.java [hunk 10]
```diff
@@ -46,4 +53,20 @@ public void testServiceDefined() {
...
+
+        WorkflowProcessInstance pi = mock(WorkflowProcessInstance.class);
+
+        when(pi.milestones()).thenReturn(null);
+        assertThat(batch.createMilestones(pi)).isNull();
+
+        when(pi.milestones()).thenReturn(emptyList());
+        assertThat(batch.createMilestones(pi)).isEmpty();
+
+        when(pi.milestones()).thenReturn(singleton(Milestone.builder().withId("id").withName("name").withStatus(Status.AVAILABLE).build()));
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply api/kogito-services/src/test/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBatchTest.java:69]: done

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java [hunk 11]
```diff
@@ -28,53 +28,31 @@
...
-import org.kie.kogito.process.NodeInstanceNotFoundException;
-import org.kie.kogito.process.NodeNotFoundException;
 import org.kie.kogito.process.Process;
-import org.kie.kogito.process.ProcessError;
-import org.kie.kogito.process.ProcessInstance;
-import org.kie.kogito.process.ProcessInstanceDuplicatedException;
-import org.kie.kogito.process.ProcessInstanceNotFoundException;
-import org.kie.kogito.process.Signal;
-import org.kie.kogito.process.WorkItem;
+import org.kie.kogito.process.*;
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:49]: yeah, fixed my settings....

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java (hunk 1 above)
Comment (cristianonicolai) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:59]: removed

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java (hunk 2 above)
Comment (cristianonicolai) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:60]: removed

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java (hunk 3 above)
Comment (cristianonicolai) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:94]: removed

File: kogito-codegen/src/test/java/org/kie/kogito/codegen/tests/PublishEventTest.java (hunk 4 above)
Comment (cristianonicolai) [reply kogito-codegen/src/test/java/org/kie/kogito/codegen/tests/PublishEventTest.java:90]: fixed

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java (hunk 7 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java:19]: done

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java (hunk 6 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:372]: Yeah, it's a good point, I had the same impression when reading the variable name _legacy_ but it seems that in reality, we're just wrapping this with a new public API. So no using `WorkflowProcessInstance` for what is needed for this PR seemed just too big of a change to me. Fact is that the core is still heavily dependent on this internal implementation. For instance, all event listeners rely on that. Perhpas is something we need to consider moving forward, if we need a big change for that before we get into 1.0 or we will simply continue with the current approach.
@mswiderski did you had any plans around that?

File: api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBody.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ruromero) [reply api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBody.java]: then why initialize it at all?

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java [hunk 12]
```diff
@@ -1089,6 +1099,46 @@ public void internalSetErrorMessage(String errorMessage) {
...
+                            .withId(uid)
+                            .withName(n.getName())
+                            .withStatus(getMilestoneStatus(uid))
+                            .build();
+                })
+                .collect(Collectors.toSet());
+    }
+
+    private <N extends Node> Stream<N> getNodesByType(Class<N> nodeClass) {
+        return Stream.of(getNodeContainer().getNodes()).filter(nodeClass::isInstance).map(nodeClass::cast);
```
Comment (mswiderski) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java:1129]: are you interested in only the first level (directly under process) or you would like to collect all milestones including those in subprocesses? If the later then you should use getNodesRecursively

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java (hunk 6 above)
Comment (mswiderski) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:381]: I don't think it will be possible anytime soon to remove the WorkflowProcessInstance and everything around it. It was not even planned to do so... what was planned and is partially done to remove the need for ksession/kruntime for it.

So yeah, the name of the variable is really misleading as it is certainly not legacy. So might be worth renaming it here.

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java (hunk 5 above)
Comment (mswiderski) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:386]: WorkflowProcessInstance is still the core of the engine so it is completely fine to add things there. This api is just a wrapper to expose it to consumers to provide more modern api and to allow independent changes on the core classes.

File: api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBody.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/ProcessInstanceEventBody.java]: because is not overriden all the time :)

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java (hunk 12 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java:1129]: hm, ok makes sense to me to get all

Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java]: done

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java (hunk 6 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java]: renamed to `processInstance` :)

**Comment by cristianonicolai** [comment]
@ruromero @mswiderski applied changes, please take another look.

File: api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/MilestoneEventBody.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        MilestoneEventBody other = (MilestoneEventBody) obj;
+        if (id == null) {
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply api/kogito-services/src/main/java/org/kie/kogito/services/event/impl/MilestoneEventBody.java]: @hasys thanks, change applied

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
jenkins please retest this
